crates/core/transaction/src/plan/build.rs

```diff
@@ -488,13 +235,149 @@ impl UnauthTransaction {
         }
 
         // Compute the binding signature and assemble the transaction.
-        let binding_signing_key = rdsa::SigningKey::from(self.synthetic_blinding_factor);
-        let auth_hash = self.inner.transaction_body.auth_hash();
+        let binding_signing_key = rdsa::SigningKey::from(synthetic_blinding_factor);
+        let auth_hash = transaction.transaction_body.auth_hash();
+
         let binding_sig = binding_signing_key.sign(rng, auth_hash.as_bytes());
         tracing::debug!(bvk = ?rdsa::VerificationKey::from(&binding_signing_key), ?auth_hash);
 
-        self.inner.binding_sig = binding_sig;
+        transaction.binding_sig = binding_sig;
+
+        Ok(transaction)
+    }
+
+    /// Build the serial transaction this plan describes.
+    /// Arguments
+    ///     - `full_viewing_key`, the [`FullViewingKey`] for the source funds;
+    ///     - `witness_data`, the [`WitnessData`] used for proving;
+    ///     -  `auth_data`, the [`AuthorizationData`] required to sign transaction.
+    pub fn build(
+        self,
+        full_viewing_key: &FullViewingKey,
+        witness_data: WitnessData,
+        auth_data: &AuthorizationData,
+    ) -> Result<Transaction> {
+        // Add the memo.
+        let mut _memo: Option<MemoCiphertext> = None;
+        let mut memo_key: Option<PayloadKey> = None;
+        if self.memo_plan.is_some() {
+            let memo_plan = self
+                .memo_plan
+                .clone()
+                .ok_or_else(|| anyhow!("missing memo_plan in TransactionPlan"))?;
+            _memo = memo_plan.memo().ok();
+            memo_key = Some(memo_plan.key);
+        }
```

Comment:
Do we need to build the memo prior to building the actions? Or can we just access the memo key from the plan?

Reply:
Also, elsewhere in the effect hash code I noticed that if the memo is missing, there's a zero payload key used instead, maybe that could work here too (cc @redshiftzero)?

Reply:
We don't need to build the memo first per se, we can just access the memo key from the `MemoPlan`, or indeed use `[0; 32]`. We only need the memo key in order to add it to any outputs.

Reply:
https://github.com/penumbra-zone/penumbra/blob/main/crates/wasm/src/planner.rs#L547 prevents us from accessing the memo key directly from the plan. We could go with @hdevalence earlier suggestion and instead take self and consume the planner. I'll play around with it.

Reply:
I don't think it does, that code is in the _planner_, which is producing the plan, not in the returned plan itself. The `build` method is on the plan and the plan is immutable, so there's not a problem here.